### ISSUE TITLE
salt-cloud libvirt: There was a profile error: virStorageVolCreateXML() argument 2 must be str or None, not bytes

### ISSUE THREAD

**Issue by gaetanquentin** [saltstack/salt#47633]
### Description of Issue/Question
Trying to create a vm:
salt-cloud -l debug -p centos7x tpl-centos7x-1

failed:
[ERROR   ] There was a profile error: virStorageVolCreateXML() argument 2 must be str or None, not bytes

debug logs:
[INFO    ] disk attributes {'device': 'disk', 'type': 'file'}
[DEBUG   ] volume: ['XMLDesc', '__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_conn', '_o', 'c_pointer', 'connect', 'delete', 'download', 'info', 'key', 'name', 'path', 'resize', 'storagePoolLookupByVolume', 'upload', 'wipe', 'wipePattern']
[DEBUG   ] Creating b'<volume>\n                    <name>centos7.0-45355366-56fe-11e8-8442-00163e244078.qcow2</name>\n                    <capacity>5368709120</capacity>\n                    <allocation>0</allocation>\n                    <target>\n                        <format type="qcow2" />\n                        <compat>1.1</compat>\n                    </target>\n                    <backingStore>\n                        <format type="qcow2" />\n                        <path>/var/lib/libvirt/storages/centos7.0.qcow2</path>\n                    </backingStore>\n                </volume>'
[INFO    ] Cleaning up after exception
[ERROR   ] There was a profile error: virStorageVolCreateXML() argument 2 must be str or None, not bytes
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/cloud/clouds/libvirt.py", line 339, in create
    clone_domain = conn.lookupByName(name)
  File "/usr/lib/python3/dist-packages/libvirt.py", line 4074, in lookupByName
    if ret is None:raise libvirtError('virDomainLookupByName() failed', conn=self)
libvirt.libvirtError: Domain not found: no domain with matching name 'tpl-centos7x-1'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/cloud/cli.py", line 281, in run
    self.config.get('names')
  File "/usr/lib/python3/dist-packages/salt/cloud/__init__.py", line 1454, in run_profile
    ret[name] = self.create(vm_)
  File "/usr/lib/python3/dist-packages/salt/cloud/__init__.py", line 1284, in create
    output = self.clouds[func](vm_)
  File "/usr/lib/python3/dist-packages/salt/cloud/clouds/libvirt.py", line 470, in create
    raise e
  File "/usr/lib/python3/dist-packages/salt/cloud/clouds/libvirt.py", line 408, in create
    new_volume = pool.createXML(create_volume_with_backing_store_xml(volume), 0)
  File "/usr/lib/python3/dist-packages/libvirt.py", line 3185, in createXML
    ret = libvirtmod.virStorageVolCreateXML(self._o, xmlDesc, flags)
TypeError: virStorageVolCreateXML() argument 2 must be str or None, not bytes

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)

/etc/salt/cloud.providers:
kvm-via-ssh:
  driver: libvirt
  url: qemu+ssh://cloud@qboxv3/system?socket=/var/run/libvirt/libvirt-sock

 /etc/salt/cloud.profiles.d/centos7x.conf:
centos7x:
  # points back at provider configuration
  provider: kvm-via-ssh
  base_domain: centos7.0
  ip_source: ip-learning
  ssh_username: xxx
  password: xxxx
  # /tmp is mounted noexec.. do workaround
  deploy_command: sh /tmp/.saltcloud/deploy.sh
  script_args: -F
  # grains to add to the minion
  grains:
    clones-are-awesome: true
  # override minion settings
  minion:
    master: salt
    master_port: 4506
  clone_strategy: quick

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)

### Versions Report
root@salt:/etc/salt/master.d# salt-cloud --versions-report
Salt Version:
            Salt: 2018.3.0
 
Dependency Versions:
 Apache Libcloud: 1.5.0
            cffi: Not Installed
        cherrypy: 3.5.0
        dateutil: 2.4.2
       docker-py: Not Installed
           gitdb: 0.6.4
       gitpython: 1.0.1
           ioflo: Not Installed
          Jinja2: 2.8
         libgit2: Not Installed
         libnacl: Not Installed
        M2Crypto: Not Installed
            Mako: Not Installed
    msgpack-pure: Not Installed
  msgpack-python: 0.4.6
    mysql-python: Not Installed
       pycparser: Not Installed
        pycrypto: 2.6.1
    pycryptodome: Not Installed
          pygit2: Not Installed
          Python: 3.5.2 (default, Nov 23 2017, 16:37:01)
    python-gnupg: Not Installed
          PyYAML: 3.11
           PyZMQ: 15.2.0
            RAET: Not Installed
           smmap: 0.9.0
         timelib: Not Installed
         Tornado: 4.2.1
             ZMQ: 4.1.4
 
System Versions:
            dist: Ubuntu 16.04 xenial
          locale: UTF-8
         machine: x86_64
         release: 4.15.0-20-generic
          system: Linux
         version: Ubuntu 16.04 xenial

**Comment by Ch3LL** [comment]
was this working previously? I wonder if this is related to the unicode work we did in the 2018.3 release.

**Comment by gaetanquentin** [comment]
@Ch3LL First time am am trying it.

**Comment by gtmanfred** [comment]
It is almost definitely unicode related.

**Comment by Ch3LL** [comment]
ping @terminalmage any ideas here?

**Comment by rklaren** [comment]
It seems looks that the `ElementTree.tostring()` API changed between python2 and python3. In python2 it returns a string in python3 it returns bytes.

**Comment by gtmanfred** [comment]
in python2, a string is bytes.  That is the biggest change between py2 and py3.

str goes from being bytes to unicode.  So it is consistently returning the same thing, always bytes.

**Comment by rklaren** [comment]
Any case the bytes trip up the call to libvirt. I have some time in the coming week to write a patch that works on both python2 and python3.

**Comment by gtmanfred** [comment]
check out salt.utils.stringutils, for some helper functions for dealing with unicode/bytes/str

**Comment by rklaren** [comment]
Update: time was a bit more elusive, and I ran into another bug while testing (a problem with a newer libvirt version in Fedora). I'm traveling next week and will probably not be able to work on it. I'll get back on it towards the end of the month. I pushed the work in progress in a branch https://github.com/rklaren/salt/tree/fix-saltcloud-libvirt-python3 so you can see if the xml conversion changes fix things.

**Comment by rklaren** [comment]
My apologies for dropping off the grid for a while, I had very little time due to a new job starting and some other issues. It looks like the unicode issue has been fixed in [this commit](https://github.com/saltstack/salt/commit/fb7885315ccc83672defab4ed0565e81ff2f3b1b#diff-b458a8419b66dbc88d8dda56ccb2e5a3) while merging in changes from 2018.3. Thank you @Vaelatern! @gtmanfred I guess this issue could be closed if it's on appropriate branches?

I'll open a separate ticket for the issue I saw on Fedora with a newer libvirt.
